main.go

```diff
@@ -474,8 +520,111 @@ func makeRequestWithTracePriority(t transport.Transport, request *transport.Requ
+	}
+	if err != nil {
+		out.Fatalf("Failed while reading stream input: %v\n", err)
+	}
+
+	req := &yarpctransport.StreamMessage{Body: ioutil.NopCloser(bytes.NewReader(msg))}
+	if err := stream.SendMessage(ctx, req); err != nil {
+		out.Fatalf("Failed while sending stream request: %v\n", err)
+	}
+	return false
+}
+
+// receiveStreamMessage receives and prints the stream message response
+// returns true if stream ends with EOF
+func receiveStreamMessage(ctx context.Context, stream *yarpctransport.ClientStream, serializer encoding.Serializer, out output) bool {
+	msg, err := stream.ReceiveMessage(ctx)
+	if err == io.EOF {
+		return true
+	}
+	if err != nil {
+		out.Fatalf("Failed while receiving stream response: %v\n", err)
+	}
+
+	bytes, err := ioutil.ReadAll(msg.Body)
+	if err != nil {
+		out.Fatalf("Failed while reading stream response: %v\n", err)
+	}
+	res, err := serializer.Response(&transport.Response{Body: bytes})
+	if err != nil {
+		out.Fatalf("Failed while serializing stream response: %v\n", err)
+	}
+
+	bs, err := json.MarshalIndent(res, "", "  ")
+	if err != nil {
+		out.Fatalf("Failed to convert map to JSON: %v\nMap: %+v\n", err, res)
+	}
+	out.Printf("%s\n\n", bs)
+	return false
+}
+
+func closeStream(ctx context.Context, stream *yarpctransport.ClientStream, out output) {
+	if err := stream.Close(ctx); err != nil {
+		out.Fatalf("Failed to close send stream: %v\n", err)
+	}
+}
+
+func makeInitialStreamRequest(out output, t transport.Transport, serializer encoding.Serializer, streamMsgReader encoding.StreamRequestReader, req *transport.Request) {
+	streamTransport, ok := t.(transport.StreamTransport)
+	if !ok {
+		out.Fatalf("Transport does not support stream calls: %q", t.Protocol())
+	}
+	streamSerializer, ok := serializer.(encoding.StreamSerializer)
+	if !ok {
+		out.Fatalf("Serializer does not support stream calls: %q", serializer.Encoding())
+	}
+
+	ctx, cancel := tchannel.NewContext(req.Timeout)
+	defer cancel()
+	ctx = makeContextWithTrace(ctx, t, req, 0)
+	stream, err := streamTransport.CallStream(ctx, req)
+	if err != nil {
+		out.Fatalf("Failed while making stream call: %v\n", err)
+	}
+
+	if streamSerializer.IsClientStreaming() && streamSerializer.IsServerStreaming() {
+		// bi-directional stream
+		for {
+			if eof := sendStreamMessage(ctx, stream, streamMsgReader, out); eof {
+				closeStream(ctx, stream, out)
+				break
+			}
+			if eof := receiveStreamMessage(ctx, stream, serializer, out); eof {
+				out.Fatalf("Received EOF while receiving bi-directional stream response: %v\n", io.EOF)
+			}
+		}
+	} else {
+		serverStreaming := streamSerializer.IsServerStreaming()
+		clientStreaming := streamSerializer.IsClientStreaming()
+		for {
+			eof := sendStreamMessage(ctx, stream, streamMsgReader, out)
+			// send message only once if the method is server streaming
+			// since server streaming method expects only one request
+			if eof || serverStreaming {
+				break
+			}
+		}
+		closeStream(ctx, stream, out)
```

Comment:
Something I am not sure to understand here. In this loop tou will first send all of the messages, close the stream and then read the response? How this is working if the stream is closed?

Reply:
Naming of `stream.Close` in YARPC is wrong 😐
Close actually closes the stream from the client side while server can still send the messages

https://github.com/yarpc/yarpc-go/blob/2cb4fe14284bfbf430f9846075fb1c61792a2527/transport/grpc/stream.go#L158

[CloseSend gRPC doc](https://godoc.org/google.golang.org/grpc#ClientStream)